File: CAML/tp1_to_3.ml
(*
###############################################################
############ AUTHOR FREVILLE TITOUAN - ID 217821 ##############
#### Feel free to use this to improve, but don't copy them ####
###############################################################

This is NOT in the COURSES of supinfo. OCAML is a child of lisp develop by INRA (french ;))
Consult documentation for more information :) http://caml.inria.fr/
CAML is a HARD TYPED LANGAGE, type constructor help to create type so you can match more things ;)
   CAML will always take the SMALLEST RESTRICTION type when prototyping the function
ALl the function are tested with SUBLIME REPL on linux (ocaml package install)
*)
(* TP 1 - La Programmation Fonctionnelle (done with a langage :O) *)
(* TP 1 - 1 - APPARTENANCE *)
(* Let rec name p1 p2 etc define a RECURSIVE function of name "name"
  using p1 p2 etc as parameters
*)
let rec find_red l_color =
  (* match X with tell to match X with values (if X == Values then do ->
    match case are introduce with |
    _ is the undefined case
  *)
  match l_color with
  (*
    l color is a string list ('a list is a basic type of OCAML)
    [] is the empty list
    t::q represent a list who have at least one element
    t is the head of the list
    q the queue :p
    you can't acces other elements than head of list (you don't have
    direct to nth elements of list in CAML)
  *)
  | [] -> false
  | t::q -> t = "red" || find_red q;;
let rec find_color l_color c =
  match l_color with
  | [] -> false
  | t::q -> t = c || find_color q c;;
(* Test the function *)
#trace find_red;;
let rest_red_1 = ["green";"green";"red"; "red"];;
let rest_red_2 = ["green";"green";"green";"blue"];;
find_red rest_red_1;;
find_red rest_red_2;;
find_color rest_red_1;;
find_color rest_red_1;;
(* TP 1 - 2 - AUTANT DE 1 QUE DE 0 *)
(* (X,Y,Z,...) is a tuple of X, Y, Z, etc ...
   access can be done with tuple.p_element
*)
let rec asmuch_naive_rec list n1 n0 =
  match list with
  | [] -> (n1,n0)
  | t::q -> if (t = 1)
              then asmuch_naive_rec q (n1+1) n0
              else asmuch_naive_rec q n1 (n0+1);;
let asmuch_naive list =
  let (a,b) = asmuch_naive_rec list 0 0 in a = b;;

let rec asmuch_naive_less_variable_rec list n =
  match list with
  | [] -> n = 0
  | t::q -> if (t = 1)
              then asmuch_naive_less_variable_rec q (n+1)
              else asmuch_naive_less_variable_rec q (n-1)
;;
let asmuch_naive_less_variable list =
  asmuch_naive_less_variable_rec list 0;;
(* Those 2 are the easiest && fastet way to solve the pb in caml
   As we can't have a quick access to the end of the list ;)
   if we want to adapt the subject to CAML :
   we have say that bits are done by having as much as possible
   1 between 0 && 0 between 1 if format like this
*)
let rec asmuch_camlformat list =
  match list with
  | [] -> true
  | t::[] -> false
  | a::b::q -> a != b && asmuch_camlformat q;;

(* TP 1 - 3 - AUTANT DE 1 QUE DE 0 *)
let rec more_naive_rec list n1 n0 =
  match list with
  | [] -> (n1,n0)
  | t::q -> if (t = 1)
              then more_naive_rec q (n1+1) n0
              else more_naive_rec q n1 (n0+1);;
let more_naive list =
  let (a,b) = more_naive_rec list 0 0 in a != b;;

let rec more_naive_less_variable_rec list n =
  match list with
  | [] -> n != 0
  | t::q -> if (t = 1)
              then more_naive_less_variable_rec q (n+1)
              else more_naive_less_variable_rec q (n-1)
;;
let more_naive_less_variable list =
  more_naive_less_variable_rec list 0;;
(* Those 2 are the easiest && fastet way to solve the pb in caml
   As we can't have a quick access to the end of the list ;)
   if we want to adapt the subject to CAML :
   we have say that bits are done by having as much as possible
   1 between 0 && 0 between 1 if format like this
*)
let rec more_camlformat list =
  match list with
  | [] -> false
  | t::[] -> true
  | a::b::q -> a = b || more_camlformat q;;

(* all the function can work with the format we described for caml
so the test case wil be like this :) *)
let test_bits_eq = [1;0;1;0;1;0];;
let test_bits_1 = [1;0;1;0;1;0;1;1;1;1];;
let test_bits_0 = [0;1;0;1;0;1;0;0];;
(* #trace more_naive;;
#trace more_naive_less_variable;;
#trace more_camlformat;; *)
asmuch_naive test_bits_eq;;
asmuch_naive test_bits_1;;
asmuch_naive test_bits_0;;
asmuch_naive_less_variable test_bits_eq;;
asmuch_naive_less_variable test_bits_1;;
asmuch_naive_less_variable test_bits_0;;
asmuch_camlformat test_bits_eq;;
asmuch_camlformat test_bits_1;;
asmuch_camlformat test_bits_0;;
(**)
more_naive test_bits_eq;;
more_naive test_bits_1;;
more_naive test_bits_0;;
more_naive_less_variable test_bits_eq;;
more_naive_less_variable test_bits_1;;
more_naive_less_variable test_bits_0;;
more_camlformat test_bits_eq;;
more_camlformat test_bits_1;;
more_camlformat test_bits_0;;

File: Python/tp1_to_3.py
###############################################################
############ AUTHOR FREVILLE TITOUAN - ID 217821 ##############
#### Feel free to use this to improve, but don't copy them ####
###############################################################

# This is the reinterpretation of tp 1 to 3 for python 
# Python tested using python 2 on Ubuntu ;) 
# FUNCTION ON LISTS WITHOUT DELETE
# @car
# @PARAM list
# @RETURN first element
def car (l):
  return l[0]
# @cdr
# @PARAM list
# @RETURN list without head
def cdr (le):
  le=list(le)
  e = car (le)
  le.remove (e)
  return le
# @last
# @PARAM list
# @RETURN value of the last element in list
def last (l):
  return l[-1]
# @butlast
# @PARAM list
# @RETURN list without last element
def butlast (le):
  le=list(le)
  le.pop(-1)
  return le
# -------------------------------------------------------------------------------------
# Labs --------------------------------------------------------------------------------
# TP 1 -  La Programmation Fonctionnelle (done with a langage :O) ---------------------
# TP 1 - 1 : Appartenance
# Function ####################################################
# @find_red
# @PARAM list
# @RETURN True if red in list, False else
def find_red (l):
  if len(l) > 0 :
    if (car(l)=='red'):
      return True
    else:
      return find_red(cdr(l))
  else:
    return False

# @find_color
# @PARAM list, c of string
# @RETURN true if c in list, false else
def find_color (l,c):
  if len(l) > 0 :
    if (car(l)==c):
      return True
    else:
      return find_color(cdr(l),c)
  else:
    return False
# #############################################################
# Test ########################################################
l1=['green','red','green']
l2=['green','blue','green','blue','green','red','red','green']
l3=['green','green','blue','blue','green','blue']
print "Result of test for find red (l1 - l2 - l3)"
print l1
print find_red(l1)
print l2
print find_red(l2)
print l3
print find_red(l3)
print "Result of test for find color (l1 - l2 - l3)"
print l1
print find_color(l1,'blue')
print l2
print find_color(l2,'blue')
print l3
print find_color(l3,'blue')
# #############################################################
# TP 1 - 1 : Autant de 1 que de 0
# As much as : naive
# @asmuch_naive
# @PARAM list compt
# @RETURN True if as much 0 as 1 , false else
def asmuch_naive (l,n):
  if len(l) > 0 :
    if (car(l)==1):
      return asmuch_naive (cdr(l),n+1)
    else:
      return asmuch_naive (cdr(l),n-1)
  else:
    return (n==0)
# As much : using last
# @asmuch_last
# @PARAM list
def asmuch_last (l):
  if len(l) > 1 :
    if (car(l)==last(l)):
      return False
    else:
      return asmuch_last (butlast(cdr(l)))
  else:
    if len(l) > 0 :
      return False
    else:
      return True
# #############################################################
# Test ########################################################
l4=[1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0]
l5=[1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0]
l6=[1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0]
print "Result as_much naive for list : "
print l4
print asmuch_naive(l4,0)
print l5
print asmuch_naive(l5,0)
print l6
print asmuch_naive(l6,0)
print "Result asmuch_last for list : "
print l4
print asmuch_last(l4)
print l5
print asmuch_last(l5)
print l6
print asmuch_last(l6)
# #############################################################